### PR TITLE
Possibility to add cookies to AVURLAsset and HTTPCookieStorage

### PR DESCRIPTION
**Pull request by e-nicolas**
This will allow download AWS CloudFront signed videos, using:

CloudFront-Key-Pair-Id
CloudFront-Policy
CloudFront-Signature

https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-signed-cookies.html

### CODE REVIEW

File: VidLoader/VidLoader/Sources/VidLoader.swift [hunk 1]
```diff
@@ -257,13 +257,20 @@ public final class VidLoader: VidLoadable {
     }
 
     private func startTask(urlAsset: AVURLAsset, streamResource: StreamResource, item: ItemInformation) {
+        addCookiesTo(urlAsset: urlAsset, headers: item.headers)
```
Comment (Cyklet) [inline VidLoader/VidLoader/Sources/VidLoader.swift:260]: Hi @e-nicolas do you have an URL that I can use to test?
I think you need to send `item.headers` in `setupResourceDelegate` to`ResourceLoader`, isn't AVAssetResourceLoaderDelegate `CookieLoader` replaced with `ResourceLoader` when `setupResourceDelegate` is called?

Comment (e-nicolas) [reply VidLoader/VidLoader/Sources/VidLoader.swift]: Hi @Cyklet, thanks for your quick reply. I saw that the headers was not necessary to be added on URLRequest header fields. I just added the cookies on HTTPCookieStorage and pass it to AVURLAsset too. This way started to download de videos properly.

On VideoListDataProvider I added the headers:

`let headers = ["CloudFront-Key-Pair-Id": "K8IK9YH5HAVIE", "CloudFront-Policy": "eyJTdGF0ZW1lbnQiOlt7IlJlc291cmNlIjoiaHR0cHM6Ly9xYS1tZWRpYS5hZnlhLmNvbS5ici82MDA4MzRiNTY3MjYwZDIxOTBmM2Y0NTkvKiIsIkNvbmRpdGlvbiI6eyJEYXRlTGVzc1RoYW4iOnsiQVdTOkVwb2NoVGltZSI6MTY1MjQ3MjY1OH19fV19", "CloudFront-Signature": "at1Q5KAYsweDfq7W3jKT9uPK51ZGaBo~32MivGBLjFTO~ZPi6qsvI8XrtdEba0b-wOZX9OaRmPS1DU0ri7tCzWJYh1harwfGRLEbc2LlOnv7POsxySDCuSEnK0AFnP~cIuisCZa-sbsZGmTWJSVh~iJqXi5JljYK66P163OVCccQqC1Pkdw-QulrCCAEyfOUEkmCva8gljpSpspHaWxJPWsBPGWVoXqbLQ0vZAm9Ybe3HS-da1Vc6voLTfxMNgH2lWHiFVpvpKN-88QIUltsLcT9fHW~pPq7jduXE8wBgEaw4E-uGAEiH1wSCbHwzeHierVjjo6eTh1ZctrEIyY0rA__" ]`

And pass it to `DownloadValues` on `startDownload` method.

I changed the video defaultURL to:
https://qa-media.afya.com.br/600834b567260d2190f3f459/master.m3u8

This signature and policy has duration of 1 day

File: VidLoader/VidLoader/Sources/Classes/ResourceLoader/SchemeHandler.swift [hunk 2]
```diff
@@ -20,18 +20,23 @@ enum SchemeType: String {
 }
 
 struct SchemeHandler: SchemeHandleable {
-    func urlAsset(with mediaURL: URL?) -> Result<AVURLAsset, ResourceLoadingError> {
+    
+    func urlAsset(with mediaURL: URL?, headers: [String: String]?) -> Result<AVURLAsset, ResourceLoadingError> {
```
Comment (Cyklet) [inline VidLoader/VidLoader/Sources/Classes/ResourceLoader/SchemeHandler.swift:24]: the `headers: [String: String]?` is unused, maybe you can create the `AVURLAssetHTTPHeaderFieldsKey` cookies array [HTTPCookie] in the scope and remove the usage of `HTTPCookieStorage.shared` saving it in a shared class when we are working with multiple threads can have side effects.

Also one more thing:
If you try to reuse headers, should cookie values also added to the video/playlist request? or it is enough only to create `AVURLAsset` with options?

I mean this place:

PlaylistLoader && ResourceLoader
        headers?.forEach {
            urlRequest.addValue($0.value, forHTTPHeaderField: $0.key)
        }

File: VidLoader/VidLoader/Sources/Utils/CookieOptionsUtils.swift [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    
+    static func insertCookieOptionsWith(domain: String?, headers: [String: String]?) {
+        guard let domain = domain else { return }
+        
+        if let headers = headers {
+            for key in headers.keys {
+                let cookie: [HTTPCookiePropertyKey : Any] = [
+                    HTTPCookiePropertyKey.domain: domain,
+                    HTTPCookiePropertyKey.path: "/",
+                    HTTPCookiePropertyKey.secure: true,
```
Comment (Cyklet) [inline VidLoader/VidLoader/Sources/Utils/CookieOptionsUtils.swift:21]: This looks project specific if cookie creation is needed only for `AVURLAsset(url: url, options: options)` then you can easily add ne property in `ItemInformation` -> something like:
`let AVURLAssetHTTPHeaderFieldsKey: [Any]?`

and pass to 

`func urlAsset(with mediaURL: URL?, headerFields: [Any]?) {
   -----///-----
     if let headerFields = headerFields {
         options["AVURLAssetHTTPHeaderFieldsKey"] = headerFields
      }
}
`

File: VidLoader/VidLoader/Sources/Classes/ResourceLoader/SchemeHandler.swift (hunk 2 above)
Comment (e-nicolas) [reply VidLoader/VidLoader/Sources/Classes/ResourceLoader/SchemeHandler.swift]: Did you managed to download the video with other way? I tried create and add only to AVURLAsset options with 2 ways below:
    
    `// 1
    func createCookieOptions(headers: [String: String]?, domain: String?) -> [HTTPCookie]? {
        guard let domain = domain else { return nil }
        
        var cookies = [HTTPCookie]()
        if let headers = headers {
            for key in headers.keys {
                let cookie: [HTTPCookiePropertyKey : Any] = [
                    HTTPCookiePropertyKey.domain: domain,
                    HTTPCookiePropertyKey.path: "/",
                    HTTPCookiePropertyKey.secure: true,
                    HTTPCookiePropertyKey.init("HttpsOnly"): true,
                    HTTPCookiePropertyKey.value: headers[key] ?? "",
                    HTTPCookiePropertyKey.name: key,
                ]
                if let httpCookie = HTTPCookie(properties: cookie) {
                    cookies.append(httpCookie)
                }
            }
        }
        return cookies
    }
    
    // 2
    func createCookieOptions(headers: [String: String]?, url: URL) -> [String: [String: String]]? {
        guard let cookie = headers else { return nil }
        var cookies = [HTTPCookie]()
        for key in cookie.keys {
            let cookieField = ["Set-Cookie": "\(key)=\(cookie[key] ?? "")"]
            let cookie = HTTPCookie.cookies(withResponseHeaderFields: cookieField, for: url)
            cookies.append(contentsOf: cookie)
        }
        let values = HTTPCookie.requestHeaderFields(with: cookies)
        return ["AVURLAssetHTTPHeaderFieldsKey": values]
    }
    `

Also adding on urlRequest headers did not work too:
   ` headers?.forEach {
        urlRequest.addValue($0.value, forHTTPHeaderField: $0.key)
      }`
      
I removed the headers parameters from SchemeHandler implementation.

**Comment by Cyklet** [comment]
Hi @e-nicolas sorry for being off for long time, how do you want to proceed with this PR? should we try to find a solution that can be used in other projects too? as I see you have made a fork so I assume you have already this changes on your side

**Comment by Cyklet** [comment]
Hi @e-nicolas I will close this PR until we don't find a scalable solution that can be used across multiple applications, we can also reopen if needed

**Pull request closed without merging**